### PR TITLE
Added bottom_up_cluster to python api

### PR DESCRIPTION
**Pull request by cchadowitz-pf**
I wanted to use bottom up clustering in a python script but realized it wasn't present (yet) so I figured I'd go ahead and implement it. Happy to tweak or adjust if need be.

### CODE REVIEW

File: tools/python/src/face_recognition.cpp [hunk 1]
```diff
@@ -230,6 +230,36 @@ class face_recognition_model_v1
...
+            dist(i2, j2) = dlib::length( first_descriptor- second_descriptor);
+            dist(j2, i2) = dist(i2, j2);
+        }
+    }
+
+    std::vector<unsigned long> labels;
+    const auto num_clusters = dlib::bottom_up_cluster(dist, labels, min_num_clusters, max_dist);
+
+    py::list clusters;
+    for (size_t i = 0; i < labels.size(); ++i) {
```
Comment (arrufat) [inline tools/python/src/face_recognition.cpp:257]:
```suggestion
    for (size_t i = 0; i < labels.size(); ++i)
    {
```

File: tools/python/src/face_recognition.cpp [hunk 2]
```diff
@@ -230,6 +230,36 @@ class face_recognition_model_v1
...
+
+    size_t num_descriptors = py::len(descriptors);
+    matrix<float> dist(static_cast<long>(num_descriptors), static_cast<long>(num_descriptors));
+
+    for (size_t i = 0; i < num_descriptors; ++i)
+    {
+        for (size_t j = i+1; j < num_descriptors; ++j)
+        {
+            const long i2 = static_cast<long>(i),
+                       j2 = static_cast<long>(j);
```
Comment (arrufat) [inline tools/python/src/face_recognition.cpp:245]:
```suggestion
            const long i2 = static_cast<long>(i);
            const long j2 = static_cast<long>(j);
```

**Comment by arrufat** [comment]
Thank you! I added a couple of minor changes to better fit the dlib coding style.
Let's wait for Davis and see what he thinks :)

File: tools/python/src/face_recognition.cpp [hunk 3]
```diff
@@ -230,6 +230,37 @@ class face_recognition_model_v1
 
 // ----------------------------------------------------------------------------------------
 
+py::list bottom_up_clustering(py::list descriptors, const int min_num_clusters, const double max_dist)
+{
+    DLIB_CASSERT(min_num_clusters > 0);
+
+    size_t num_descriptors = py::len(descriptors);
+    matrix<float> dist(static_cast<long>(num_descriptors), static_cast<long>(num_descriptors));
```
Comment (davisking) [inline tools/python/src/face_recognition.cpp:238]: Did you get a compiler warning without adding these `static_casts`?  They shouldn't be needed.

**Comment by davisking** [comment]
Sweet, thanks. Looks good.  I'll merge once the tests run :)

File: tools/python/src/face_recognition.cpp (hunk 3 above)
Comment (cchadowitz-pf) [reply tools/python/src/face_recognition.cpp:238]: Not here, no, but I reused this code from another place where I used this code pattern and must have had a reason to do so there. Happy to remove it here if you'd like though.

**Pull request merged**